opm/autodiff/FlowMain.hpp

```diff
@@ -255,6 +256,77 @@ namespace Opm
 
 
 
+        /// helper function checking that _one_ coordinate differs by 1
+        bool ijk_are_adj(int i1, int j1, int k1, int i2, int j2, int k2) {
+            return ((i1-i2)*(i1-i2) +
+                    (j1-j2)*(j1-j2) +
+                    (k1-k2)*(k1-k2)) == 1;
+        }
+
+        std::array<int, 3> getIJK(size_t g, const Grid& grid) {
+            using namespace UgGridHelpers;
+            std::array<int, 3> r;
+            const int * dimens = cartDims(grid);
+            size_t m_nx = dimens[0];
+            size_t m_ny = dimens[1];
+            size_t m_nz = dimens[2];
+
+            int i = g % m_nx;  g /= m_nx;
+            int j = g % m_ny;  g /= m_ny;
+            int k = g % m_nz;
+
+            r[0] = i; r[1] = j; r[2] = k;
+            return r;
+        }
+
+        /// checks cartesian adjacency of global indices g1 and g2
+        bool cartesianAdjacent(size_t g1, size_t g2, const Grid& grid) {
+
+            auto cell1 = getIJK(g1, grid);
+            auto cell2 = getIJK(g2, grid);
+            return ijk_are_adj(cell1[0], cell1[1], cell1[2],
+                               cell2[0], cell2[1], cell2[2]);
+        }
+
+        /**
+         * Get the NNC structure of the given grid---that is, cell adjacencies
+         * beyond Cartesian neighborhoods.  The transmissivity is always given as
+         * -1d (unless trans vector is provided) hence NNC "structure".
+         */
+        NNC nncStructure(const Grid& grid, const std::vector<double>& trans = std::vector<double>()) {
+            // we use numFaces, numCells, cell2Faces, globalCell from UgGridHelpers
+            using namespace UgGridHelpers;
+            NNC nnc;
+
+            typedef std::vector<int> nb; // neighbors
+            size_t num_faces = numFaces(grid);
+            if (trans.size() > 0 && trans.size() != num_faces)
+                throw std::logic_error(
+                            "non-empty trans vector with wrong dimension.");
+
+            std::vector<nb> faces(num_faces); // a "2 by f matrix" face2cells
+            for (int cell = 0; cell < numCells(grid); cell++) { // active cell idx
+                for (auto face : cell2Faces(grid)[cell]) {
```

Comment:
`face` is actually just an integer (or size_t) - whereas here It could look like an object of some sort; maybe call the variable `face_nr` or just use the type explicitly.